File: Semester 3/DSA/EXP1-Implementing-Stack-and-Queue-using-Switch case/Stack-using-array/Stack.c
#include <stdio.h>


//Defining Max Element of Stack
#define max 5 

//Declearing Stack
int Stack[max];

//Declearing Top as Null or -1 (So that when incremented it will get value 0 i.e. array stack first element)
int top=-1;

void Push()
{   
    //Checking If Stack is Full 
    if(top==max-1)
        printf("Stack Overflow");
    
    //Taking Element From user ,incrementing the index of array and assigning taken value to the index i.e. Top
    else
    {
        int x;
        printf("Enter The Number to Be Pushed: ");
        scanf("%d",&x);
        top++;
        Stack[top]=x;
    }
}

void Pop()
{   
    //Checking if Stack is Empty 
    if(top==-1)
        printf("Stack UnderFlow");

    /*
     As Top point to last element of array and Stack is LIFO, 
     we show last element in Array using Stack,
     and reduce the Top 
    */
    else
    {
        printf("Number Poped is %d \n",Stack[top]);
        top--;
    }
}

void Peek()
{
    printf("TopMost Element in Stack is %d \n",Stack[top]);
}

void display()
{   
    //using for loop to display the Stack, incrementing i till top as top is the position of the topmost element in stack
    for (int i = 0; i <= top; i++)
    {
        printf("%d  ",Stack[i]);
    }
    
}

int main()
{
    //Declearing integer choice for switch case and loop for while loop
    int choice,Loop=1;

    //while loop is used to make the switch case run again and again till loop integer is change other than 1
    while(Loop==1)
    {
         printf(" ----Menu--- \n");
         printf("| 1.Push    |\n");
         printf("| 2.Pop     |\n");
         printf("| 3.Peek    |\n");
         printf("| 4.Display |\n");
         printf(" ----------- \n");
         printf("Enter the Choice: ");
         scanf("%d",&choice);
    
   
        switch(choice)
        {
             case 1:
               Push();
             break;

             case 2:
                Pop();
             break;

             case 3:
              Peek();
             break;

             case 4:
                display();
             break;

             default:
                printf("Please! Enter Valid Choice");

            
        }

        printf("Do you want to Continue(Enter 1 to Continue or Any other Number to Exit: ");
        scanf("%d",&Loop);
        
    }

            return 0;
}

File: Semester 4/Python Lab/EXP5/stack.py
"""Write a Python program to implement a stack using an array."""

import array

# Initialize an empty array as a stack
stack = array.array('i')

# Push operation
def push(val):
    stack.append(val)

# Pop operation
def pop():
    if len(stack) == 0:
        print("Stack is empty")
    else:
        return stack.pop()

# Display stack
def display():
    print(stack)


while(True):
    print("\n1. Push\n2. Pop\n3. Display\n4. Exit")
    choice = int(input("Enter your choice: "))

    if choice == 1:
        val = int(input("Enter the value to push: "))
        push(val)
    elif choice == 2:
        popped = pop()
        if popped is not None:
            print(f"Popped element: {popped}")
    elif choice == 3:
        display()
    elif choice == 4:
        break
    else:
        print("Invalid choice. Try again.")